File: bin/json2code.py
#!/usr/bin/env python

import os
import sys
import json
import jinja2

def camelize(value):
  return "".join(x.capitalize() if x else '_' for x in value.split("_"))

def classize(value):
	return value[:1].upper() + value[1:]

def instantize(value):
	return value[:1].lower() + value[1:]

JINJA_ENVIRONMENT = jinja2.Environment(
  loader=jinja2.FileSystemLoader("platforms"),
  extensions=['jinja2.ext.autoescape'],
  trim_blocks=True,
  lstrip_blocks=True,
  autoescape=True)

JINJA_ENVIRONMENT.filters['classize'] = classize
JINJA_ENVIRONMENT.filters['instantize'] = instantize
JINJA_ENVIRONMENT.filters['camelize'] = camelize

schema_path = sys.argv[1]
prefix = sys.argv[2]
package = sys.argv[3]

schema = json.load(open(schema_path))

allClasses = {}

def parseClass(root, name):
	one_of = root.get('oneOf', {})
	if one_of:
		parseOneOf(one_of, name)
	else:
		properties = root.get('properties', {})
		for property_name in properties.iterkeys():
			property = properties[property_name]
			# Dereference references to objects and enums
			if property.get("$ref"):
				property['type'] = get_reference(property.get("$ref"))
				del property['$ref']

		allClasses[name] = {
			'kind': 'class',
			'properties': properties,
			'required': root.get('required', [])
		}

def parseEnum(root, name):
	allClasses[name] = {
		'kind': 'enum',
		'enum': root.get('enum', []),
		'type': root.get('type', 'string')
	}

def get_reference(reference):
	if reference[:2] != "#/":
		raise RuntimeError('Only internal absolute references are allowed at this time')

	refComponents = reference[2:].split('/')
	element = schema
	while len(refComponents) > 0:
		key = refComponents.pop(0)
		element = element.get(key)

	if not key in allClasses:
		if element.get('type') == 'object':
			parseClass(element, key)
		elif element.get('enum'):
			parseEnum(element, key)
		else:
			raise RuntimeError('Only enums are allowed by reference at this time')

	return key

def parseOneOf(one_of, name):
	possibles = []
	for item in one_of:
		if item.get('$ref'):
			ref = get_reference(item.get('$ref'))
			possibles.append(ref)
		else:
			raise RuntimeError('oneOf only supports referenced definitions at this time')

	allClasses[name] = {
		'kind': 'oneOf',
		'possibles': possibles
	}

def generate(platform, name, definition):
	name = classize(name)
	print "Generating %s %s ..." % (platform, name)

	for template_file in os.listdir('platforms/' + platform + '/object'):
		template = JINJA_ENVIRONMENT.get_template(platform + '/object/' + template_file)

		result = template.render({
			'prefix': prefix,
			'package': package,
			'name': name,
			'kind': definition.get('kind', 'class'),
			'required': definition.get('required', []),
			'properties': definition.get('properties', {}),
			'possibles': definition.get('possibles', []),
			'enum': definition.get('enum', []),
			'oneOf': definition.get('oneOf', False),
			'allClasses': allClasses
		})

		folder = 'output/' + platform
		if platform == 'android':
			folder = folder  + '/' + '/'.join(package.split('.'))

		if not os.path.exists(folder):
			os.makedirs(folder)

		_, ext = os.path.splitext(template_file)
		if platform == 'android':
			with open(folder + '/' + prefix + name + ext, 'w') as file:
				file.write(result)
		else:
			with open(folder + '/' + (prefix if platform == 'ios' else '') + name + ext, 'w') as file:
				file.write(result)
		
		

parseClass(schema, schema.get('title'))

platforms = os.listdir('platforms')
for platform in platforms:
	for name, definition in allClasses.iteritems():
		generate(platform, name, definition)

File: json2code.js
'use strict';

const fs = require('fs');
const mkdirp = require('mkdirp');
const nunjucks = require('nunjucks');
const path = require('path');

nunjucks.installJinjaCompat();
const env = nunjucks.configure(path.join(__dirname, 'platforms'), {
  trimBlocks: true,
  lstripBlocks: true,
});

const [schemaPath, prefix, pkg] = process.argv.slice(2);

const schema = JSON.parse(fs.readFileSync(path.resolve(schemaPath)));

function classize(value) {
  return value.substr(0, 1).toUpperCase() + value.substr(1);
}

env.addFilter('classize', classize);

function instantize(value) {
  return value.substr(0, 1).toLowerCase() + value.substr(1);
}
env.addFilter('instantize', instantize);

function camelize(value) {
  const x = value.split('_').map((element) => {
    if (element) {
      return classize(element.toLowerCase());
    }
    return '_';
  }).join('');
  console.log('camelize', value, x);
  return x;
}
env.addFilter('camelize', camelize);

const allClasses = {};

function parseClass(root, name) {
  if (root.oneOf) {
    parseOneOf(root.oneOf, name);
  } else {
    const properties = Object.keys(root.properties).sort().reduce((previous, propertyName) => {
      let property = root.properties[propertyName];
      // Dereference references to objects and enums
      if (property.$ref) {
        property = Object.assign({}, property, {
          type: getReference(property.$ref),
        });
      }
      return previous.concat([[
        propertyName,
        property,
      ]]);
    }, []);

    // const properties = root.properties;
    // Object.keys(properties).forEach((propertyName) => {
    //   const property = properties[propertyName];
    //   if (property.$ref) {
    //     property.type = getReference(property.$ref);
    //     delete property.$ref;
    //   }
    // });

    allClasses[name] = {
      kind: 'class',
      properties,
      required: root.required || [],
    };
    console.log('Adding parsed class', allClasses[name]);
  }
}

function parseEnum(root, name) {
  allClasses[name] = {
    kind: 'enum',
    enum: root.enum || [],
    type: root.type,
  };
}

function getReference(reference) {
  if (reference.substr(0, 2) !== '#/') {
    throw new Error('Only internal absolute references are allowed at this time');
  }

  const refComponents = reference.substr(2).split('/');
  let element = schema;
  let key;
  while (refComponents.length > 0) {
    key = refComponents.shift();
    element = element[key];
  }

  if (!allClasses[key]) {
    if (element.type === 'object') {
      parseClass(element, key);
    } else if (element.enum) {
      parseEnum(element, key);
    } else {
      throw new Error('Only enums are allowed by reference at this time');
    }
  }

  return key;
}

function parseOneOf(oneOf, name) {
  const possibles = [];
  oneOf.forEach((item) => {
    if (item.$ref) {
      const ref = getReference(item.$ref);
      possibles.push(ref);
    } else {
      throw new Error('oneOf only supports referenced definitions at this time');
    }
  });

  allClasses[name] = {
    kind: 'oneOf',
    possibles,
  };
}

function generate(platform, name, definition) {
  name = classize(name);
  console.log(`Generating ${platform} ${name} ...`);

  const templateFiles = fs.readdirSync(path.join(__dirname, 'platforms', platform, 'object'));

  templateFiles.forEach((templateFile) => {
    console.log(templateFile);
    const templateFilePath = path.join(platform, 'object', templateFile);
    const result = env.render(templateFilePath, {
      prefix,
      package: pkg,
      name,
      kind: definition.kind || 'class',
      required: definition.required,
      properties: definition.properties,
      possibles: definition.possibles,
      enum: definition.enum,
      oneOf: definition.oneOf || false,
      allClasses,
    });

    let folder = path.join('output', platform);
    if (platform === 'android') {
      folder = path.join(folder, path.join(...pkg.split('.')));
    }

    mkdirp.sync(folder);

    const ext = path.extname(templateFile);
    let outputFileName = `${name}${ext}`;
    if (platform === 'ios' || platform === 'android') {
      outputFileName = `${prefix}${name}${ext}`;
    }
    console.log('writing to', path.join(folder, outputFileName));
    fs.writeFileSync(path.join(folder, outputFileName), result);
  });
}

console.log(schema);

parseClass(schema, schema.title);

console.log(allClasses);

const platforms = fs.readdirSync(path.join(__dirname, 'platforms'));

platforms.forEach((platform) => {
  console.log('platform', platform);
  Object.keys(allClasses).forEach((name) => {
    const definition = allClasses[name];
    generate(platform, name, definition);
  });
});
